### ISSUE TITLE
Rule proposal: no-typos

### ISSUE THREAD

**Issue by dwhieb** [eslint/eslint#10762]
**Please describe what the rule should do:**

The rule should check for typos, based on a blacklist of misspelled words. The user would provide this list in the options for the plugin. Another option would be that the blacklist is case-sensitive.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Get teh language Object <-- "the" is misspelled
const langauge = getLanguage(); // "language" is misspelled

const costs = dikstrasAlgorithm(graph); // "Dijkstra's" is misspelled
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a very common use case for most developers, especially when working on domain-specific projects. While ESLint already catches certain typos that cause syntax errors, most scripts are written for specific domains, such as math, linguistics, library science, product management, etc. As such, each developer's scripts will likely contain many repetitions of words that are common for those domains (e.g. "language" will appear many times in scripts written for and by linguists). If developers find themselves frequently making typos, they can blacklist those words, catching the typo before it causes problems in things like variable naming in the code. Typos are especially problematic because IDEs will autocomplete words based on existing ones, meaning that your IDE might autocomplete the same typo several times. These cases are subtle because the spelling differences are often small, and this problem wouldn't violate ESLint's `no-unused-vars` rule, so you'd have no way of knowing you had made the typo until your script throws an error.

**Comment by ilyavolodin** [comment]
A lot of those typos would be flagged by a combination of `no-unused-vars`, `no-undef` and so on. I'm with @kaicataldo this seems like a good rule for a plugin, but I'm not sure it could be implemented cleanly enough for the core.
